Resolve the Savd Confirmation bug for Issue #83 and add Javadoc comments to new method

WhiteHsu opened the pull request:
In order to resolve the Issue #83 , I have the following changes in this pull request:
- I create the new "SaveConfirmDialog" class under the "latexstudio.editor.menu" package.
![pic01](https://cloud.githubusercontent.com/assets/7831418/14430567/ad624e58-0034-11e6-9953-d239f550b804.png)
- In the "actionPerformed" method of the "SaveFile" class, after user selects a file in the file choose dialog, it'll call the "confirmFileSave" method from the new "SaveConfirmDialog" class to confirm if there's a file overwriting.
![pic02](https://cloud.githubusercontent.com/assets/7831418/14430577/b442428c-0034-11e6-9a62-d15cefc5bc7c.png)
- In the "SaveAs" class, I also rewrite its original confirmation codes by calling the "confirmFileSave" method from the new "SaveConfirmDialog" class.
![pic03](https://cloud.githubusercontent.com/assets/7831418/14430585/ba49329e-0034-11e6-9931-6c35e3ea8d3c.png)
- Besides, I find there's tiny issue in the original conditional expression of the while loop for rewriting confirmation, so I update the conditional logic as well.
![pic04](https://cloud.githubusercontent.com/assets/7831418/14430589/c0bff93c-0034-11e6-8877-d4091472dbe0.png)

Because this's my first time to create pull request, please kindly help review and provide your valuable feedback. Any advice will be very appreciated. Thank you very much. :-)

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/menu/SaveConfirmDialog.java:
```diff
@@ -0,0 +1,35 @@
+ * @return the selected file object
+ */
+public class SaveConfirmDialog {
+    public static File confirmFileSave(int reply, File file, File currentFile) {
```
I think this method could be placed in the FileChooserService. After all, it can only be used with FileChooserService anyways.

sebbrudzinski replied:
It seems like "reply" shouldn't be a method parameter but this method's variable.

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/menu/SaveConfirmDialog.java:
```diff
@@ -0,0 +1,35 @@
+ */
+public class SaveConfirmDialog {
+    public static File confirmFileSave(int reply, File file, File currentFile) {
+        while (file != null && file.exists() && file != currentFile && reply == JOptionPane.NO_OPTION) {            
```
I think the file comparison (file != currentFile) should use equals method.

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/menu/SaveConfirmDialog.java:
```diff
@@ -0,0 +1,35 @@
+        while (file != null && file.exists() && file != currentFile && reply == JOptionPane.NO_OPTION) {            
+            reply = JOptionPane.showConfirmDialog(null,
+                            file.getAbsoluteFile() + " already exists. Do you want to overwrite it?",
+                            "File already exists.",
```
No dots in titles

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/menu/SaveFile.java:
```diff
@@ -45,6 +46,10 @@ public void actionPerformed(ActionEvent e) {
         if (file == null) file = FileChooserService.getSelectedFile("tex", "TeX files", DialogType.SAVE, true);
+        
+        //Confirm if to overwrite an exsiting file.
+        file = SaveConfirmDialog.confirmFileSave(JOptionPane.NO_OPTION, file, etc.getCurrentFile());
```
I don't like that we must first call "getSelectedFile" and then additionaly "confirmFileSave". Could we have a method that merges the logic of the two above? It could be named "getFileWithConfirmation".

sebbrudzinski commented:
@WhiteHsu Thanks for the pull request! It doesn't look bad overall, however I've left some feedback inline.

WhiteHsu commented:
@sebbrudzinski 
Thanks for your valuable feedback!! I've updated codes based on your review comments, you can find most your reviews have been implemented in the commit now. However, I have question about your comments on the following review feedback:
![pic01](https://cloud.githubusercontent.com/assets/7831418/14467044/120d7d00-010c-11e6-8738-89c0cca7fd3b.png)

I'm not quite sure which should be the best way to merge these two methods together. So far my way to merge them is to create another method named "getFileWithConfirmation" in the "FileChooserService" class to call the methods "getSelectedFile" and "confirmFileSave":
![pic02](https://cloud.githubusercontent.com/assets/7831418/14467122/6e0c1d8c-010c-11e6-85c6-ad38d63299e3.png)

Because the "SaveFile" and "SaveAs" have different behaviors here, I also updated more codes and add another Enum object "SAVEAS" in "DialogType", then the Confirmation Dialog can also have different behaviors accordingly. Please kindly help suggest if there's another better way:
![pic03](https://cloud.githubusercontent.com/assets/7831418/14467223/da0d93ee-010c-11e6-9b43-f5b34ded4f91.png)

In addition, in the While loop under the method "confirmFileSave", after much study I think the "Not Equal" logic of [file != currentFile] should be still kept. Here's the case: If the user clicks the "Save" button after he has already saved file previously, because the "file" is equal to "currentFile" in this case, the While loop logic doesn't meet (which has Not Equal logic), then the Confirm Dialog doesn't need to show out, and the file will be saved again directly. I've tried to amend codes with "Equal" logic but found there'll be issue on the Save function. So I still keep this "Not Equal" logic. 

Please kindly help provide your suggestions and opinions. Any advice will be very appreciated. Thank you very much.  :-)

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/files/FileChooserService.java:
```diff
@@ -102,4 +107,41 @@ public static File getSelectedFile(String extension, String description, DialogT
+        if(file == null || type == DialogType.SAVEAS)
+            file = FileChooserService.getSelectedFile(extension, description, type, fixExtension);
+        
+        file = FileChooserService.confirmFileSave(file, currentFile);
```
We can drop the "FileChooserService" now from those method calls.

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/files/FileChooserService.java:
```diff
@@ -102,4 +107,41 @@ public static File getSelectedFile(String extension, String description, DialogT
+    * @param currentFIle current file saved in Editor Top Component
+    * @return the selected file object
+    */
+    public static File confirmFileSave(File file, File currentFile) {
```
This should be private.

sebbrudzinski commented:
@WhiteHsu They way you have implemented it was pretty much what I had in mind. 
Concerning the equal/not equal, I didn't really mean changing not equal to equal. I meant changing the way the equality is checked. [file != currentFile] checks if the two class instances are the exact same thing. This MAY work for now but doesn't really check what we want to check. We want to check if the two files are not equal (their paths are not the same) and this can be achieved using equals method. In this case, we should negate the equals, to act as not equal, like [!file.equals(currentFile)]. 
More on comparing objects in Java: http://stackoverflow.com/questions/13387742/compare-two-objects-with-equals-and-operator

WhiteHsu commented:
@sebbrudzinski 
Many thanks for your feedback. I've revised codes to remove the stupid mistakes on "private" and "FileChooserService", which should not exist at all..... :-(

I've also updated the While loop logic based on your comment: To compare file objects with "! equal" instead of "!=". Your totally correct, this's surely the better way for object comparison. Many thanks for your valuable suggestion, which is so helpful.  :-)

Please kindly help review if there's any other part needing to refine as well. Thank you very much.

WhiteHsu commented:
@sebbrudzinski 
I'm sorry that I'm not familiar to the code review process. I see you change label with "code review +1" on this PR, does it mean that my code revisions are passed? Or do we need more others to review codes as well before code merge? 

Please kindly help suggest and guide me for the next. This project is so interesting, and I'm wondering if I can work on other tasks then. Thank you very much.  :-)

sebbrudzinski commented:
@WhiteHsu The code looks good to me, but I (or someone else) still need to test it. Once the testing is finished, this change will be merged.

sebbrudzinski commented on OpenLaTeXStudio/EditorModule/src/latexstudio/editor/files/FileChooserService.java:
```diff
@@ -102,4 +107,47 @@ public static File getSelectedFile(String extension, String description, DialogT
+    */
+    private static File confirmFileSave(File file, File currentFile) {
+        int reply = JOptionPane.NO_OPTION;
+        while (file != null && file.exists() && !file.equals(currentFile) && reply == JOptionPane.NO_OPTION) {            
```
We should change the [reply == NO_OPTION] to [reply != YES_OPTION]. Currently, if I close the confirmation window with X, it saves the file. In my opinion closing the confirmation dialog (and any other operation other than clicking YES) should result in aborting saving.

sebbrudzinski commented:
I've found one issue during testing (which wasn't introduced by you, but since it's a small fix, I think it's a good time to do it)

WhiteHsu commented:
@sebbrudzinski 
Thanks for your comments. I've revised codes based on you suggestion. Please kindly help review. :-)
